core/types/serialization.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/doltgresql/server/types"
+	"github.com/dolthub/doltgresql/utils"
+)
+
+// Serialize returns the TypeCollection as a byte slice.
+// If the TypeCollection is nil, then this returns a nil slice.
+func (pgs *TypeCollection) Serialize(ctx context.Context) ([]byte, error) {
```

Comment:
The serialization/deserialization functions seem like good candidates for unit tests. 

Also... we may want to plan for the future and include a serialization version here. I would not be surprised if we needed to change the data we store at some point in the future, and without serialization version metadata stored with the data, that can get pretty tough. We're in an alpha period right now, so there's not pressure for backwards compatibility, but it seems like it'd be easy to add a ser/deser format version identifier now and have it set up for the future. 

@Hydrocharged may have an opinion or thoughts on this, too.

Reply:
It's included here on line 39 (as of the writing of this comment):
```go
writer.VariableUint(0) // Version
```

We've also got the version check in the `Deserialization` function:
```go
if version != 0 {
	return nil, fmt.Errorf("version %d of types is not supported, please upgrade the server", version)
}
```
You must have accidentally missed it XD

Reply:
Oh very nice! My bad; thanks for pointing it out. I must have totally glossed over that on Friday afternoon when I was reviewing. 🤷‍♂️